Use stripe-mock for tests

ob-stripe opened the pull request:
Updated version of #384.

ob-stripe commented:
I've updated the PR to add the ability to verify requests, and also stub requests (see the functional test for `FileUpload` for an example with stubbed requests).

I think we have a good base to start porting all tests to stripe-mock. To summarize, for each model class, we want:
- a functional test class that tests all API methods (using stripe-mock whenever possible, or stubbed requests otherwise)
- a model test class that tests deserialization. If the model has at least one expandable attribute, there should be two tests: one without expanding any attributes, and one expanding all attributes

If we want to be exhaustive, we should also have deserialization tests for collection classes.

@brandur-stripe @remi-stripe Any thoughts or questions? Anything I missed before we start porting all tests?

brandur-stripe commented:
> I think we have a good base to start porting all tests to stripe-mock. To summarize, for each model class, we want:
>
> * a functional test class that tests all API methods (using stripe-mock whenever possible, or stubbed requests otherwise)
> * a model test class that tests deserialization. If the model has at least one expandable attribute, there should be two tests: one without expanding any attributes, and one expanding all attributes
If we want to be exhaustive, we should also have deserialization tests for collection classes.

Yep, that sounds absolutely perfect. I love that it'll get us a pretty exhaustive test of all possible expansions in a pretty direct way too.

> If we want to be exhaustive, we should also have deserialization tests for collection classes.

IMO, it'd be fine just to make sure that this works for just one collection, but it's up to you :)

> @brandur-stripe @remi-stripe Any thoughts or questions? Anything I missed before we start porting all tests?

Just that this is going to be a monstrous amount of work :) The existing diff looks great.

ob-stripe commented on src/test/java/com/stripe/functional/AccountTest.java:
```diff
@@ -1,7 +1,11 @@
+import com.stripe.net.APIResource;
 import com.stripe.model.*;
 import junit.framework.Assert;
 import org.junit.Test;
```
Kind of a nit, but at some point in the future I'd like to enforce Google's Java style in the project. Can you reorder the import like in the other tests?

1. All static imports first
2. All imports in aphabetical order (`com.` / `org.` prefixes are ignored)
3. Separate import "groups" with a new line (in most cases the groups will be java, junit, then stripe)
4. Don't use `.*`, explicitly import all the classes you need.

ob-stripe commented on src/test/java/com/stripe/functional/AccountTest.java:
```diff
@@ -12,22 +16,132 @@
-		accountParams.put("limit", 3);
-		AccountCollection accountCollection = Account.all(accountParams);
+	@Test
+	public void testAccountUpdate() throws StripeException {
```
Rename to `testUpdate()`

ob-stripe commented on src/test/java/com/stripe/functional/AccountTest.java:
```diff
@@ -12,22 +16,132 @@
+	}
+
+	@Test
+	public void testAccountReject() throws StripeException {
```
Rename to `testReject()`

ob-stripe commented on src/test/java/com/stripe/model/AccountTest.java:
```diff
@@ -21,14 +19,12 @@
 import static org.mockito.Mockito.*;
```
Same comment about imports as above.

ob-stripe commented on src/test/java/com/stripe/model/AccountTest.java:
```diff
@@ -21,14 +19,12 @@
 	@Test
 	public void testDeserialize() throws StripeException, IOException {
-		String json = getResourceAsString("account.json");
+		String json = getResourceAsString("/api_fixtures/account.json");
```
Rather than deserializing from an embedded JSON fixture, I'd rather we use stripe-mock whenever possible.

remi-stripe replied:
I disagree partially because stripe-mock return non Custom accounts which won't deserialize a lot of the more complex sub-elements like `legal_entity` and `external_accounts`

remi-stripe replied:
I'm happy to do both but that deserialization via stripe-mock happens in most of the other tests since we retrieve an account before reject/update/etc.

ob-stripe commented on src/test/java/com/stripe/model/AccountTest.java:
```diff
@@ -21,14 +19,12 @@
 		Account acc = APIResource.GSON.fromJson(json, Account.class);
 
-		assertEquals("acct_1032D82eZvKYlo2C", acc.getId());
+		assertEquals("acct_1234", acc.getId());
```
I don't think it's necessary to check the value of every field (especially if you're using stripe-mock instead of an embedded fixture file). Mostly what we're testing here is that `fromJson` does not raise an exception.

In the other deserialization tests, I used just two asserts: `assertNotNull(resource)` + `assertNotNul(resource.getId())` to check for the unlikely case where deserialization would not raise an exception but return `null`, or return an instance with `null` values in attributes.

For models with submodels (e.g. `AccountPayoutSchedule`), we should create a separate deserialization test for the submodel, e.g. in `model/AccountPayoutScheduleTest.java`:
```java
String accountData = getFixture("/v1/accounts/acct_123");
String data = getDataAt(accountData, "payout_schedule");
AccountPayoutSchedule object = APIResource.GSON.fromJson(data, AccountPayoutSchedule.class);
assertNotNull(object);
assertNotNull(object.getInterval());
```

(cf. `ChargeOutcomeTest.java` in the first commit)

remi-stripe replied:
That seems pretty painful to maintain for the Account resource and all its sub-resources and various versions.

ob-stripe replied:
They are separate classes in the code, so they should be tested separately. It will certainly be a bit painful to cover all models (mostly due to the fact that our coverage is pretty spotty right now), but once that's done, maintenance should be minimal.

remi-stripe commented:
@ob-stripe PTAL. I addressed all your comments (I think) and changed the Account model tests to only use stripe-mock's fixture. Then I moved LegalEntityTest to the new framework

ob-stripe commented:
👍  Nice, looks great! Thanks for also deleting the old fixtures that are no longer used after you updated the tests.

remi-stripe commented on src/test/java/com/stripe/functional/IdempotentTest.java:
```diff
@@ -1,68 +0,0 @@
-package com.stripe.functional;
```
@ob-stripe We need to look into moving those tests to a different level to make sure we are actively verifying this works but it's the wrong way to test idempotency

remi-stripe replied:
bumping this just in case

brandur-stripe commented on src/test/java/com/stripe/functional/StripeResponseTest.java:
```diff
@@ -35,12 +35,13 @@ public void testResponseIncluded() throws
     assertEquals(200, resp.code());
-    assertEquals(idempotencyKey, resp.idempotencyKey());
+    // TODO: ob add this to stripe-mock
+    // assertEquals(idempotencyKey, resp.idempotencyKey());
```
Well, it still doesn't do anything too smart with idempotency keys, but I just patched stripe-mock to reflect an idempotency key that you sent back to you like the real API does. Patch in https://github.com/stripe/stripe-mock/pull/77 and release in 0.17.0.

remi-stripe replied:
Damn you're so fast, I had barely started the code lol.

ob-stripe commented on src/test/java/com/stripe/model/ExternalAccountTest.java:
```diff
@@ -48,52 +37,44 @@ public void testUnknownExternalAccountRetrieval() throws StripeException, IOExce
 
-    stubNetwork(ExternalAccount.class,
-        "{\"id\": \"extacct_123\", \"object\": \"unknown_external_account\"}");
+    assertEquals(true, ea instanceof ExternalAccount);
```
You can use `assertTrue` here.

ob-stripe commented:
Okay, I think this is ~ready.

Coverage numbers are still fairly low primarily because of the large number of untested deprecated methods (I've tried configuring the Cobertura plugin to ignore those, but it doesn't seem to be working too well) and some setter methods (which should also be excluded from coverage but I haven't looked into it yet).

Note that there is one non-test change in this PR: I added a new method `Stripe.overrideUploadBase()` in order to be able to use stripe-mock to test file uploads. We already had similar methods for the API and for OAuth anyway.

r? @brandur-stripe @remi-stripe

remi-stripe commented on src/test/java/com/stripe/functional/ApplicationFeeTest.java:
```diff
@@ -1,22 +1,59 @@
+    assertNotNull(resources);
+    verifyRequest(
+        APIResource.RequestMethod.GET,
+        String.format("/v1/application_fees")
```
should we verify that params are passed?

remi-stripe replied:
but there are params here right?

remi-stripe commented on src/test/java/com/stripe/functional/BalanceTransactionTest.java:
```diff
@@ -1,34 +1,43 @@
+    assertNotNull(resources);
+    verifyRequest(
+        APIResource.RequestMethod.GET,
+        String.format("/v1/balance/history")
```
should we verify that params are sent? (won't repeat it for all retrieve, might be worth fixing all)

remi-stripe replied:
same as above, we are explicitly passing `limit` but not verifying it.

remi-stripe commented on src/test/java/com/stripe/functional/BankAccountTest.java:
```diff
@@ -0,0 +1,166 @@
+
+  @Test
+  public void testRetrieve() throws StripeException {
+    final Customer customer = Customer.retrieve(CUSTOMER_ID);
```
why do we have `final` here and not on any of the other examples? Should we fix them all? (annoying I know but consistency is king)

ob-stripe replied:
Immutability is good, so ideally all variables that can be declared as `final` _should_ be. I think there's a linter rule that require that variables whose declaration is more than X lines away from their first use be declared as final, which is what prompted me to start using it in some places, and then thought that we might as well use it everywhere.

In practice, I don't think it's worth stressing about too much.

remi-stripe commented on src/test/java/com/stripe/functional/BankAccountTest.java:
```diff
@@ -0,0 +1,166 @@
+        String.format("/v1/customers/%s/sources", customer.getId()),
+        params,
+        ExternalAccountCollection.class,
+        stubbedCollection.toJson()
```
I don't really understand this one. Why not a fixture that is a list already? I think that's what I did for other tests.

ob-stripe replied:
Because I'd be mildly annoyed to have a fixture file with a bank account, and a different fixture file with the same bank account wrapped in a list.

Hopefully stripe-mock will be able to handle this case sooner rather than later and those fixtures will become unnecessary anyway.

remi-stripe replied:
FWIW I did that for other resources though.

remi-stripe commented on src/test/java/com/stripe/functional/CardTest.java:
```diff
@@ -1,20 +1,150 @@
+        getResourceAsString("/api_fixtures/card.json"), Card.class);
+    card.setCustomer(customer.getId());
+
+    return card;
```
you likely want a comment to explain that this is because stripe-mock is mean to you.

remi-stripe commented on src/test/java/com/stripe/functional/CardTest.java:
```diff
@@ -1,20 +1,150 @@
+        String.format("/v1/customers/%s/sources", customer.getId()),
+        params,
+        Card.class,
+        getResourceAsString("/api_fixtures/card.json")
```
why not `getCardFixture`?

ob-stripe replied:
`stubRequest` wants a `Class` and a `String` containing the JSON. `getCardFixture()` returns a `Card` instance. We could pass `getCardFixture().toJson()` but that seems a bit redunant (we'd deserialize the JSON into a `Card` only to reserialize it to JSON).

remi-stripe commented on src/test/java/com/stripe/functional/CardTest.java:
```diff
@@ -1,20 +1,150 @@
+        String.format("/v1/customers/%s/sources/%s", customer.getId(), CARD_ID),
+        null,
+        Card.class,
+        getResourceAsString("/api_fixtures/card.json")
```
same as above

ob-stripe replied:
same as above :)

remi-stripe commented on src/test/java/com/stripe/functional/CardTest.java:
```diff
@@ -1,20 +1,150 @@
+        params,
+        ExternalAccountCollection.class,
+        stubbedCollection.toJson()
+    );
```
same as with the bank account, why not a real fixture?

ob-stripe replied:
same as above

remi-stripe commented on src/test/java/com/stripe/functional/CouponTest.java:
```diff
@@ -1,128 +1,89 @@
-    customer.deleteDiscount();
-    assertNull(Customer.retrieve(customer.getId()).getDiscount());
-  }
+    resource.delete();
```
should we test `CouponDeleted` like you did for Card (if so, applies to all delete)

ob-stripe replied:
Probably. The important part of the test is really `verifyRequest`, but might as well be consistent.

remi-stripe commented on src/test/java/com/stripe/functional/FeeRefundTest.java:
```diff
@@ -0,0 +1,85 @@
+  public void testList() throws StripeException {
+    final ApplicationFee fee = ApplicationFee.retrieve(FEE_ID);
+
+    final FeeRefundCollection reversals = fee.getRefunds().list(null);
```
rename the variable to `resources` instead (bad copy paste I think)

ob-stripe replied:
Ah yeah, this one should be `refunds`. (I understand that using `resource` / `resources` everywhere makes the initial effort of porting/writing tests easier, but I prefer explicit names.)

remi-stripe replied:
> I understand that using resource / resources everywhere makes the initial effort of porting/writing tests easier, but I prefer explicit names.

I disagree, that avoids the copy/paste fails we have in most of our libraries. But would be great to be consistent since we already went through so much trouble to make everything consistent.

remi-stripe commented on src/test/java/com/stripe/functional/LoginLinkTest.java:
```diff
@@ -0,0 +1,44 @@
+        getResourceAsString("/api_fixtures/login_link.json")
+    );
+
+    final LoginLink link = account.getLoginLinks().create();
```
I wonder if it'd be cleaner to have an account fixture with the `login_links` collection instead

ob-stripe replied:
No strong opinion on this -- I don't think it matters too much how the fixtures are generated as long as it doesn't impede readability.

remi-stripe commented on src/test/java/com/stripe/functional/OrderReturnTest.java:
```diff
@@ -0,0 +1,39 @@
+    final OrderReturnCollection orderReturns = OrderReturn.list(null);
+
+    assertNotNull(orderReturns);
+    assertEquals(1, orderReturns.getData().size());
```
should we do this assert on other list endpoints (I don't think I've seen it yet)
also why do we expect `1` when we're not even passing a limit?

remi-stripe commented on src/test/java/com/stripe/functional/OrderTest.java:
```diff
@@ -0,0 +1,106 @@
+
+  @Test
+  public void testList() throws StripeException {
+    final OrderCollection orders = Order.list(null);
```
should be called `resources`

remi-stripe commented on src/test/java/com/stripe/functional/OrderTest.java:
```diff
@@ -0,0 +1,106 @@
+    Map<String, Object> params = new HashMap<String, Object>();
+    params.put("currency", "usd");
+
+    final Order order = Order.create(params);
```
should be called `resource`

remi-stripe commented on src/test/java/com/stripe/functional/PlanTest.java:
```diff
@@ -2,205 +2,91 @@
-    Customer customer = createDefaultCustomerWithPlan(plan);
-    assertEquals(customer.getSubscriptions().getData().get(0).getPlan().getId(), plan.getId());
-  }
+    Plan plan = Plan.create(params);
```
should be called `resource`

remi-stripe commented on src/test/java/com/stripe/functional/ProductTest.java:
```diff
@@ -0,0 +1,106 @@
+    params.put("package_dimensions", packageDimensions);
+    params.put("type", "good");
+
+    final Product product = Product.create(params);
```
should be called `resource`

remi-stripe commented on src/test/java/com/stripe/functional/ReversalTest.java:
```diff
@@ -0,0 +1,85 @@
+  public void testList() throws StripeException {
+    final Transfer transfer = Transfer.retrieve(TRANSFER_ID);
+
+    final TransferReversalCollection reversals = transfer.getReversals().list(null);
```
should be called `resources`

remi-stripe commented on src/test/java/com/stripe/functional/SKUTest.java:
```diff
@@ -0,0 +1,103 @@
+import com.stripe.BaseStripeTest;
+import com.stripe.exception.StripeException;
+import com.stripe.model.DeletedSKU;
+import com.stripe.model.SKU;
```
flagging for next major version, are we renaming to Sku? 😱

ob-stripe replied:
We can make an exception for SKU.

remi-stripe commented on src/test/java/com/stripe/functional/SKUTest.java:
```diff
@@ -0,0 +1,103 @@
+
+  @Test
+  public void testList() throws StripeException {
+    SKUCollection skus = SKU.list(null);
```
should be called `resources`

remi-stripe commented on src/test/java/com/stripe/functional/SKUTest.java:
```diff
@@ -0,0 +1,103 @@
+    params.put("product", "prod_123");
+    params.put("image", "http://example.com/foo.png");
+
+    SKU sku = SKU.create(params);
```
should be called `resource`

remi-stripe commented on src/test/java/com/stripe/functional/SourceTest.java:
```diff
@@ -1,40 +1,137 @@
+    params.put("currency", "usd");
+    params.put("owner", owner);
+
+    Source source = Source.create(params);
```
should be called resource

remi-stripe commented on src/test/java/com/stripe/functional/SubscriptionItemTest.java:
```diff
@@ -1,106 +1,92 @@
+    params.put("subscription", "cus_123");
+    params.put("quantity", 99);
+
+    SubscriptionItem subscriptionItem = SubscriptionItem.create(params);
```
should be called `resource`

remi-stripe commented on src/test/java/com/stripe/functional/SubscriptionTest.java:
```diff
@@ -3,220 +3,108 @@
-    sub = sub.cancel(null);
-    assertNotNull(sub.getCanceledAt());
-  }
+    Subscription subscription = Subscription.create(params);
```
should be called resource.

remi-stripe commented on src/test/java/com/stripe/functional/TimeoutTest.java:
```diff
@@ -1,23 +0,0 @@
-package com.stripe.functional;
```
restarting that I think we should find a way to test this. Maybe stripe-mock could have an endpoint that explicitly waits X seconds to respond (as a parameter) and we'd put a timeout that is X - 1 ?

remi-stripe commented on src/test/java/com/stripe/functional/TopupTest.java:
```diff
@@ -0,0 +1,80 @@
+    params.put("description", "creating a topup");
+    params.put("statement_descriptor", "topup");
+
+    Topup topup = Topup.create(params);
```
should be named `resource`

remi-stripe commented on src/test/java/com/stripe/functional/TokenTest.java:
```diff
@@ -1,82 +1,49 @@
+    Map<String, Object> params = new HashMap<String, Object>();
+    params.put("card", card);
+
+    Token token = Token.create(params);
```
should be named `resource`

remi-stripe commented on src/test/java/com/stripe/functional/TransferTest.java:
```diff
@@ -1,103 +1,78 @@
-    } catch (InvalidRequestException ire) {
-      // do nothing
-    }
+    Transfer transfer = Transfer.create(params);
```
should be named `resource`

remi-stripe commented on src/test/java/com/stripe/functional/UsageRecordTest.java:
```diff
@@ -1,65 +1,39 @@
+    // UsageRecord.create() has a non-standard behavior: it uses the `subscription_item` element
+    // in the params map to form the endpoint's URL. We need to remove it from the map because
+    // it won't be sent as a URL-encoded parameter.
+    params.remove("subscription_item");
```
flagging that we should fix this in the next major version and move back to a parameter for the id

remi-stripe commented on src/test/java/com/stripe/model/AccountTest.java:
```diff
@@ -1,195 +1,19 @@
+    String data = getFixture("/v1/accounts/acct_123");
+    Account resource = APIResource.GSON.fromJson(data, Account.class);
+    assertNotNull(resource);
+    assertNotNull(resource.getId());
```
we likely want specific edge-case as fixtures to ensure deserialization works as needed.

remi-stripe commented on src/test/java/com/stripe/model/BalanceTransactionTest.java:
```diff
@@ -1,32 +1,30 @@
-    assertNotNull(dp);
-    assertEquals("dp_1AF5AmKac2qaDcCZHvOjyrDo", dp.getId());
+  public void testDeserializeExpansions() throws Exception {
+    // TODO: Figure out why stripe-mock does not expand source when asked
```
likely worth looking/flagging and then removing the TODO

remi-stripe commented on src/test/java/com/stripe/model/ChargeOutcomeTest.java:
```diff
@@ -1,51 +1,22 @@
 import org.junit.Test;
 
+// stripe-mock doesn't return an outcome, disable the test for now
+@Ignore
```
we should write a test for this

remi-stripe commented on src/test/java/com/stripe/model/DeserializerTest.java:
```diff
@@ -1,139 +0,0 @@
-package com.stripe.model;
```
assuming it's normal that this is entirely removed?

remi-stripe commented on src/test/java/com/stripe/model/EphemeralKeyTest.java:
```diff
@@ -1,104 +1,26 @@
+    EphemeralKey resource = APIResource.GSON.fromJson(json, EphemeralKey.class);
+    assertNotNull(resource);
+    assertEquals("ephkey_123", resource.getId());
+    assertEquals("ephemeral_key", resource.getObject());
```
why not use stripe-mock?

ob-stripe replied:
Because ephemeral keys can only be created, not retrieved, and require special parameters for the API version. Since this test is specifically for the deserialization, I didn't want to introduce additional complexity by using stripe-mock here.

remi-stripe commented on src/test/java/com/stripe/model/LegalEntityTest.java:
```diff
@@ -1,39 +1,19 @@
-    assertEquals(null, le.getVerification().getDetails());
+    assertNotNull(le);
+    // TODO: Figure out how to test various versions of legal_entity
+    // such as with additional_owners, with expanded verification doc, etc.
```
flagging there's a big/important todo here

remi-stripe commented on src/test/java/com/stripe/model/UsageRecordTest.java:
```diff
@@ -1,61 +1,19 @@
+    String data = getResourceAsString("/api_fixtures/usage_record.json");
+    UsageRecord resource = APIResource.GSON.fromJson(data, UsageRecord.class);
+    assertNotNull(resource);
+    assertNotNull(resource.getId());
```
why not use stripe-mock?

ob-stripe replied:
Like `EphemeralKey`, `UsageRecord` cannot be retrieved.

remi-stripe commented on src/test/resources/api_fixtures/file_upload.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "id": "file_1BrE64KCFFPkgtRhvXK81rmT",
```
random but let's not put a real id!

remi-stripe commented on src/test/resources/api_fixtures/card.json:
```diff
@@ -0,0 +1,26 @@
+{
+  "id": "card_1CaO622eZvKYlo2C22sr7okI",
```
card_123

remi-stripe commented on src/test/resources/api_fixtures/file_upload_list.json:
```diff
@@ -0,0 +1,17 @@
+  "object": "list",
+  "data": [
+    {
+      "id": "file_1BrE64KCFFPkgtRhvXK81rmT",
```
file_123

remi-stripe commented on src/test/resources/api_fixtures/source_detached.json:
```diff
@@ -0,0 +1,37 @@
+{
+  "id": "src_1BoJ2NKrZ43eBVAb7Qj7bnQL",
```
src_123

remi-stripe commented on src/test/resources/api_fixtures/source_detached.json:
```diff
@@ -0,0 +1,37 @@
+      "routing_number": "110000000",
+      "swift_code": "TSTEZ122"
+  },
+  "client_secret": "src_client_secret_CCiTVxSUWh1dzjv2qPyYuz55",
```
src_..._123

ob-stripe commented on src/main/java/com/stripe/Stripe.java:
```diff
@@ -10,9 +10,9 @@
-  public static final String UPLOAD_API_BASE = "https://uploads.stripe.com";
   public static final String LIVE_API_BASE = "https://api.stripe.com";
   public static final String CONNECT_API_BASE = "https://connect.stripe.com";
+  public static final String UPLOAD_API_BASE = "https://files.stripe.com";
```
Yes, both `files.stripe.com` and `uploads.stripe.com` point to the same IP.

ob-stripe commented on src/test/java/com/stripe/functional/AccountTest.java:
```diff
@@ -1,36 +1,145 @@
+    );
+  }
+
+  // TODO: Decide if sub-resources tests should live here.
```
So I kind of decided that already by adding `CardTest` and `BankAccountTest`, though currently they only test the customer sources case. We can add a few more tests specifically for `instanceURL` though.

brandur-stripe commented:
Can you maybe kick this back to me once Remi's feedback is satisfied and I'll give it one more pass over? Just looking at the general structure, I suspect that I won't have many qualms — the broad strokes all look totally right. I'm sure we'll find more to fix, but we can easily offload that to other PRs. This one is likely to become stale quickly.

ob-stripe commented:
Alright, I probably missed more than a few things, but:

- all tests now use `final` whenever possible
- the tests should be a lot more consistent
- params are tested whenever we send params
- I removed unnecessary fixtures
- all fixtures now use fake IDs (`smth_123`)
- I added a `charge_outcome.json` fixture file and enabled the `ChargeOutcome` deserialization test
- I added some janky but working tests for connection and read timeouts in `RequestOptionsTest`

ptal @remi-stripe

ob-stripe commented on src/test/java/com/stripe/functional/RequestOptionsTest.java:
```diff
@@ -0,0 +1,77 @@
+  @Test
+  public void testConnectTimeout() throws IOException, StripeException {
+    // Kind of terrible, but let's test connection timeouts with a random external IP.
+    Stripe.overrideApiBase("https://254.254.254.254");
```
Yeah, actually, this wasn't a great idea (and also kind of rude to whoever owns this IP). I changed it to `api.stripe.com`. Hopefully Travis' servers are more than 1 millisecond away from Stripe's 😐

brandur-stripe commented on src/test/java/com/stripe/BaseStripeTest.java:
```diff
@@ -1,170 +1,481 @@
+    }
+
+    Mockito
+        .doReturn(APIResource.GSON.fromJson(response, clazz))
```
Is this indentation right? Seems weird that the style is two spaces everywhere else, but then four spaces when chaining method calls like this.

brandur-stripe commented:
Wow, WHAT A DIFF! Thanks for doing so much work.

Okay, admittedly I only took a very high-level pass to see if I could spot anything untoward, but this LGTM. Very much looking forward to getting this in.

ptal @ob-stripe
